File: 프로그래머스/3/42895. N으로 표현/N으로 표현.py
def solution(N, number):
    dp = [set() for i in range(9)]
    dp[0] = 0
    dp[1] = {N}
    dp[2] = {N+N, N*11, N-N, N*N, N/N}

# dp[i] -> dp[i-1] +-*/ dp[1], dp[i-2] dp[2] ... 반대로 한번 더, int("N"*i) 
    for i in range(1, 9):
        dp[i].add(int(str(N) * i))
        for j in range(1, i):
            for x in dp[j]:
                for y in dp[i-j]:
                    dp[i].add(x + y)
                    dp[i].add(x - y)
                    dp[i].add(x * y)
                    if y != 0:
                        dp[i].add(x // y)
        if number in dp[i]:
            return i
    
    return -1

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.java
import java.util.*;

class Solution {
    public int solution(int N, int number) {
        HashSet<Integer>[] dp = new HashSet[9];
        for (int i = 1; i <= 8; i++) {
            dp[i] = new HashSet<>();
        }
        dp[1].add(N);
        dp[2].add(N * N);
        dp[2].add(N + N);
        dp[2].add(N * 11);
        dp[2].add(N / N);

        for (int i = 1; i <= 8; i++) {
            StringBuilder sb = new StringBuilder();
            for (int k = 0; k < i; k++) {
                sb.append(N);
            }
            dp[i].add(Integer.parseInt(sb.toString()));
            
            for (int j = 1; j < i; j++) {
                for (int x : dp[j]) {
                    for (int y : dp[i - j]) {
                        dp[i].add(x + y);
                        dp[i].add(x - y);
                        dp[i].add(x * y);
                        if (y != 0) {
                            dp[i].add(x / y);
                        }
                    }
                }
            }
            if (dp[i].contains(number)) {
                return i;
            }
        }        
        return -1;
    }
}